File: HotDesk_SPICE/Library/Library.Common/Component/Generator.vb
Imports System.Data
Imports System.IO
Imports System.Web.UI

Public Class Generator
    Implements ICollection(Of FieldSet)

#Region "Properties"
    Private _data As DataTable
    ''' <summary>
    ''' Data Source
    ''' </summary>
    Public Property Data() As DataTable
        Get
            Return _data
        End Get
        Set(ByVal value As DataTable)
            _data = value
        End Set
    End Property

    Private _setting As New List(Of FieldSet)
    ''' <summary>
    ''' Field Setting Collection
    ''' </summary>
    Public ReadOnly Property Setting() As List(Of FieldSet)
        Get
            Return _setting
        End Get
    End Property
#End Region

    ''' <summary>
    ''' initial the data
    ''' </summary>
    Public Sub New(ByVal data As DataTable)
        Me._data = data
        Me._setting = New List(Of FieldSet)
    End Sub

    Public Sub AddField(ByVal Field As String, ByVal Title As String, ByVal Type As EnumLib.DataType)
        Me._setting.Add(New FieldSet(Field, Title, Type))
    End Sub

    Public Sub Add(ByVal item As FieldSet) Implements System.Collections.Generic.ICollection(Of FieldSet).Add
        Me._setting.Add(item)
    End Sub

    ''' <summary>
    ''' Generate the HTML
    ''' </summary>
    Public Overrides Function ToString() As String
        Using _sw As New StringWriter
            Using _htw As New HtmlTextWriter(_sw)
                Dim _dg As New System.Web.UI.WebControls.DataGrid
                Dim _counter As Integer = -1

                ' Field setting
                For Each Itm As FieldSet In Me._setting
                    _counter += 1
                    Me._data.Columns(Itm.Field).SetOrdinal(_counter)
                    Me._data.Columns(Itm.Field).AllowDBNull = True
                    Me._data.Columns(Itm.Field).ColumnName = Itm.Title
                Next

                ' Remove unwanted column
                _counter += 1
                Dim _totalColumnCount As Int16 = Me.Data.Columns.Count - 1
                For temp As Integer = _counter To _totalColumnCount
                    Me.Data.Columns.RemoveAt(_counter)
                Next

                _dg.DataSource = Me.Data
                _dg.DataBind()
                _dg.RenderControl(_htw)

                Return _sw.ToString
            End Using
        End Using
    End Function

    Public Sub Clear() Implements System.Collections.Generic.ICollection(Of FieldSet).Clear
        Me._setting.Clear()
    End Sub

    Public Function Contains(ByVal item As FieldSet) As Boolean Implements System.Collections.Generic.ICollection(Of FieldSet).Contains
        Return Me._setting.Contains(item)
    End Function

    Public Sub CopyTo(ByVal array() As FieldSet, ByVal arrayIndex As Integer) Implements System.Collections.Generic.ICollection(Of FieldSet).CopyTo
        Me._setting.CopyTo(array, arrayIndex)
    End Sub

    Public ReadOnly Property Count() As Integer Implements System.Collections.Generic.ICollection(Of FieldSet).Count
        Get
            Return Me._setting.Count
        End Get
    End Property

    Public ReadOnly Property IsReadOnly() As Boolean Implements System.Collections.Generic.ICollection(Of FieldSet).IsReadOnly
        Get
            Return False
        End Get
    End Property

    Public Function Remove(ByVal item As FieldSet) As Boolean Implements System.Collections.Generic.ICollection(Of FieldSet).Remove
        Return Me._setting.Remove(item)
    End Function

    Public Function IndexOf(ByVal item As FieldSet) As Integer
        Return Me._setting.IndexOf(item)
    End Function

    Public Sub Insert(ByVal index As Integer, ByVal item As FieldSet)
        Me._setting.Insert(index, item)
    End Sub

    Default Public Property Item(ByVal index As Integer) As FieldSet
        Get
            Return Me._setting.Item(index)
        End Get
        Set(ByVal value As FieldSet)
            Me._setting.Item(index) = value
        End Set
    End Property

    Public Sub RemoveAt(ByVal index As Integer)
        Me._setting.RemoveAt(index)
    End Sub

    Public Function GetEnumerator() As System.Collections.Generic.IEnumerator(Of FieldSet) Implements System.Collections.Generic.IEnumerable(Of FieldSet).GetEnumerator
        Return Me._setting.GetEnumerator
    End Function

    Public Function GetEnumerator1() As System.Collections.IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
        Return Me._setting.GetEnumerator
    End Function
End Class

File: HotDesk_PRAI/Library/Library.Common/Component/Generator.cs
using System.Data;
using System.IO;
using System.Web.UI;
using System.Collections.Generic;

public class Generator : ICollection<FieldSet>
{
    #region Properties
    private DataTable _data;
    /// <summary>
    /// Data Source
    /// </summary>
    public DataTable Data
    {
        get { return _data; }
        set { _data = value; }
    }

    private List<FieldSet> _setting = new List<FieldSet>();
    /// <summary>
    /// Field Setting Collection
    /// </summary>
    public List<FieldSet> Setting
    {
        get { return _setting; }
    }
    #endregion

    /// <summary>
    /// initial the data
    /// </summary>
    public Generator(DataTable data)
    {
        this._data = data;
        this._setting = new List<FieldSet>();
    }

    public void AddField(string field, string title, EnumLib.DataType type)
    {
        this._setting.Add(new FieldSet(field, title, type));
    }

    public void Add(FieldSet item)
    {
        this._setting.Add(item);
    }

    /// <summary>
    /// Generate the HTML
    /// </summary>
    public override string ToString()
    {
        using (var _sw = new StringWriter())
        {
            using (var _htw = new HtmlTextWriter(_sw))
            {
                var _dg = new System.Web.UI.WebControls.DataGrid();
                int _counter = -1;

                // Field setting
                foreach (FieldSet itm in this._setting)
                {
                    _counter += 1;
                    this._data.Columns[itm.Field].SetOrdinal(_counter);
                    this._data.Columns[itm.Field].AllowDBNull = true;
                    this._data.Columns[itm.Field].ColumnName = itm.Title;
                }

                // Remove unwanted column
                _counter += 1;
                short _totalColumnCount = (short)(this.Data.Columns.Count - 1);
                for (int temp = _counter; temp <= _totalColumnCount; temp++)
                {
                    this.Data.Columns.RemoveAt(_counter);
                }

                _dg.DataSource = this.Data;
                _dg.DataBind();
                _dg.RenderControl(_htw);

                return _sw.ToString();
            }
        }
    }

    public void Clear()
    {
        this._setting.Clear();
    }

    public bool Contains(FieldSet item)
    {
        return this._setting.Contains(item);
    }

    public void CopyTo(FieldSet[] array, int arrayIndex)
    {
        this._setting.CopyTo(array, arrayIndex);
    }

    public int Count
    {
        get { return this._setting.Count; }
    }

    public bool IsReadOnly
    {
        get { return false; }
    }

    public bool Remove(FieldSet item)
    {
        return this._setting.Remove(item);
    }

    public int IndexOf(FieldSet item)
    {
        return this._setting.IndexOf(item);
    }

    public void Insert(int index, FieldSet item)
    {
        this._setting.Insert(index, item);
    }

    public FieldSet this[int index]
    {
        get { return this._setting[index]; }
        set { this._setting[index] = value; }
    }

    public void RemoveAt(int index)
    {
        this._setting.RemoveAt(index);
    }

    public IEnumerator<FieldSet> GetEnumerator()
    {
        return this._setting.GetEnumerator();
    }

    System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
    {
        return this._setting.GetEnumerator();
    }
}
